# TASK 4.5

# Consistency checks on products dataframe

# This script contains the following points:
## 01. Importing libraries
## 02. Importing Data
## 03. Data Consistency Checks:
## - Mixed-Type Data 
## - Finding Missing Values
## - Addressing Missing Values
## - Duplicates
## 04. Export data

# Step 1.

## 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data 

In [2]:
path = r'C:\Users\Sanja\Documents\08-2020 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 03. Data Consistency Checks

### Mixed-Type Data - TEST

In [5]:
# Create a dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


### Finding Missing Values

In [9]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True] 

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
df_prods.shape

(49693, 5)

### Duplicates

In [16]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_dups.shape

(5, 5)

In [19]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

### Tidying Up and Exporting Changes

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Consistency checks on orders wrangled dataframe

## Step 2.  Interpret the output of the df.describe() function, share in a markdown cell whether anything about the data looks off or should be investigated further.
### Tip: Keep an eye on min and max values!

In [22]:
# Dropping the first column of the dataFrame for the orders wrangled , because it was exported with the index column
df_ords = df_ords.iloc[:,1:]

In [23]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 6 columns]

In [24]:
# Change of the data type of the columns from a number to a string to ignore when calculating descriptive statistics
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [25]:
# Change of the data type of the columns from a number to a string to ignore when calculating descriptive statistics
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [26]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_last_order  
count           3.214874e+06  
mean            1.111484e+01  
std             9.206737e+00  
min             0.000000e+00  
25%             4.000000e+00  
50%             7.000000e+00  
75%             1.500000e+01  
max             3.000000e+01

## A2. 
## Investigating the frequency table, in particular the max and min section, i didn't  find anyting strange about the data.
### - order_number,  result in a minimum of 0 and a maximum of 100
### - orders_day_of_week, result in a minimum of 0 and a maximum of 6, for a total of exactly 7 value.
### - oder_hour_of_day, result in a minimum of 0 and a maximum of 23, for a total of exactly 24 value.
### - days_sonce_last_order, result in a minimum of 0 and a maximum of 30, for a total of exactly 31 value (Nr of days in a month).


## Step 3. Check for mixed-type data in your df_ords dataframe.

In [27]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## A3. There are no mixed data type in the df_ords dataframe

## Step 4.  If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

## Step 5. Run a check for missing values in your df_ords dataframe

In [28]:
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [29]:
df_nan_ords = df_ords[df_ords['days_since_last_order'].isnull() == True] 

In [30]:
df_nan_ords

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
11       2168274       2             1                   2                 11   
26       1374495       3             1                   1                 14   
39       3343014       4             1                   6                 11   
45       2717275       5             1                   3                 12   
...          ...     ...           ...                 ...                ...   
3420930   969311  206205             1                   4                 12   
3420934  3189322  206206             1                   3                 18   
3421002  2166133  206207             1                   6                 19   
3421019  2227043  206208             1                   1                 15   
3421069  3154581  206209             1                   3                 11   

         days_since_last_order  
0                          NaN  
11                         NaN  
26                         NaN  
39                         NaN  
45                         NaN  
...                        ...  
3420930                    NaN  
3420934                    NaN  
3421002                    NaN  
3421019                    NaN  
3421069                    NaN  

[206209 rows x 6 columns]

In [31]:
# Change of the data type of the columns from a number to a integer to ignore when calculating descriptive statistics
df_ords['user_id'] = df_ords['user_id'].astype('int64')

In [32]:
max_value = df_ords['user_id'].max()

In [33]:
max_value

206209

## A5. There are 206,209 missing values in the 'days_since_last_order' column. For all rows with missing values in this column, the 'order_number' is 1. This corresponds to 206,209 unique 'user_ids,' ranging from 1 to 206,209. This observation implies that every user's first order has a missing value in the 'days_since_last_order' column. This is logical because, for the first order, we cannot calculate the days since the last order since there is no prior order to reference.

## Step 6. Address the missing values using an appropriate method

## A6. While it might seem like there are missing values, in reality, the data in the 'days_since_last_order' column holds meaningful information. It helps in identifying new customers and their first orders. 

## Step 7. Run a check for duplicate values in your df_ords data

In [34]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [35]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## Step 8. Address the duplicates using an appropriate method.

## A8. No duplicates found

In [40]:
df_ords.shape

(3421083, 6)

## Step 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [36]:
# Change of the data type of the columns from a number to a integer to ignore when calculating descriptive statistics
df_ords['order_id'] = df_ords['order_id'].astype('int64')

In [37]:
df_ords['order_id'].dtype

dtype('int64')

In [38]:
df_ords['user_id'].dtype

dtype('int64')

In [39]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))